build_test/lib/src/test_bootstrap_builder.dart

```diff
@@ -76,3 +107,34 @@ class TestBootstrapBuilder extends Builder {
     }
   }
 }
+
+/// Manages a cache of [Configuration] per package, to avoid duplicating work
+/// across build steps.
+///
+/// Can safely be used across the build as configuration is invalidated by its
+/// digest.
+class _ConfigLoader {
+  _ConfigLoader._();
+
+  static final instance = _ConfigLoader._();
+
+  final _configByPackage = <String, Future<Configuration>>{};
```

Comment:
When would we have a different package? I don't think we'd let you read this for anything but the root package since it's not in `lib/` right?

Reply:
In theory you could have tests under lib or something, or in some other build system scenario maybe you have more than one. This just future proofs things and doesn't have any assumptions about that.

Reply:
The package comes from the primary input ID, so it works for any test under any package.

Reply:
Will `reader.canRead` always return false for any `AssetId(package, 'dart_test.yaml')` where `package` is not the root package, because we don't include other sources in the asset graph?

https://github.com/dart-lang/build/blob/93de086d54c5e70c7c504fc5ab1130df81d5d741/build_runner_core/lib/src/generate/build_definition.dart#L170-L173

Reply:
Yes it will - in build_runner. These are theoretically general builders though and that could change. We could also want to plug this into some other build system some day which does not have that same restriction.

Reply:
I don't think we should be writing this for a theoretical build system, but on the other hand it doesn't seem like it makes the code any more complicated since we don't know the root package anyway... 🤷

Reply:
right, I would otherwise be writing some assert that I only ever see the same package - which wouldn't end up being any more complicated and this way it just isn't an issue 🤷